guides/create_an_app/3-persistence.md

```diff
@@ -1,5 +1,103 @@
+
+```
+$ mix ecto.gen.migration CreateMessages
+```
+
+Add to the change function, following code:
+```
+    create table(:messages) do
+      add :name, :string
+      add :message, :string
+
+      timestamps()
+    end
+```
+
+`:messages` - name for table in database.
+
+`:name`, `:message` - columns in the database table.
+
+`:string` - field type. More on DB types you can find: [Ecto types](https://hexdocs.pm/ecto/Ecto.Schema.html#module-primitive-types)
+
+Then run the migration:
+
+```
+$ mix ecto.migrate
+```
+
+**Coach:** Explain, what is changeset? Why do we validate fields?
+
+We need to define a module, that will map the database fields into an Elixir struct. Create a file, called: `message.ex` in the directory: `lib/your_app_name/conversations`.
+
+```Elixir
+defmodule WorkshopsApp.Conversations.Message do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "messages" do
+    field :message, :string
+    field :name, :string
+
+    timestamps()
+  end
+
+  @doc false
+  def changeset(message, attrs) do
+    message
+    |> cast(attrs, [:name, :message])
+    |> validate_required([:name, :message])
+  end
+end
+```
+
+Let's create a module, where we will create a message. Create a file, called: `conversations.ex` in the directory: `lib/your_app_name/`.
+
+```
+defmodule WorkshopsApp.Conversations do
+  import Ecto.Query, warn: false
+  alias WorkshopsApp.Repo
+
+  alias WorkshopsApp.Conversations.Message
+
+  def create_message(attrs \\ %{}) do
+    %Message{}
+    |> Message.changeset(attrs)
+    |> Repo.insert()
+  end
+end
+```
+
+# Creating our messages in the Database
+
+Next, we need to save the messages as they come in. If you are not familiar with programming this is create in the CRUD app methodology. In `lib/slackir_web/channels/random_channel.ex`, we can add a call to do it, so the `handle_in` function looks like this:
+
+```Elixir
+def handle_in("shout", payload, socket) do
+  Slackir.Conversations.create_message(payload)
+  broadcast socket, "shout", payload
+  {:noreply, socket}
+end
+```
+
+# Refactor
+
+The above code is alright, but if we want to scale our application we will start to have a problem as all the saves to the database will slow down the request cycle and keep us from handling the traffic efficiently. Elixir has some built in tools for dealing with exactly this situation. Specifically, `Kernel.spawn`, which takes a module, a function, and arguments to pass to the function. So we’ll update our handler, and add a function:
+
+```Elixir
+def handle_in("shout", payload, socket) do
+  spawn(Slackir.Conversations, :create_message, [payload])
+  broadcast! socket, "shout", payload
+  {:noreply, socket}
+end
+```
+
+What this does is spawn an elixir process to do the save, outside of the request cycle. Since we don’t need to see the results of the save to broadcast the message, there’s no reason to wait for it.
 
```

Comment:
In the end we could add an alternative approach (shortcut) with context generator and some description:
```
$ mix phx.gen.context Conversations Message messages name:string message:string
Explanation on that command:

Context: Conversations - The context is an Elixir module that serves as an boundary for the given resources like messages. A context often holds many related resources.
Schema: Message - The schema is responsible for mapping the database fields into an Elixir struct.
Schema table: messages - The schema table is just a schema plural name for table in database.
Fields name:string and message:string- The fields are columns in database table where message is a filed name and string is field type in message:string.
```
WDYT?

Reply:
I'll add short info about it. 👍